<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Thang V Pham" />

<meta name="date" content="2021-01-24" />

<title>betabinom: The Beta-Binomial Test for Count Data</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">betabinom: The Beta-Binomial Test for Count
Data</h1>
<h4 class="author">Thang V Pham</h4>
<h4 class="date">2021-01-24</h4>



<p>A number of technology platforms in proteomics and genomics produce
count data for quantitative analysis. In proteomics, the number of MS/MS
events observed for a protein in a mass spectrometry experiment has been
shown to correlate strongly with the proteinâ€™s abundance. In genomics,
next-generation sequencing technologies use read counts as a measure of
the abundance of the target transcripts. The R package
<em>betabinom</em> contains functions for statistical significance
analysis of count data for both paired and unpaired designs.</p>
<div id="quick-start" class="section level2">
<h2>Quick start</h2>
<p>The user needs to install the package once, most likely by entering
<code>install.packages(&quot;betabinom&quot;)</code> in the R console, and load
the package before use.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(betabinom)</span></code></pre></div>
<p>The following are three complete examples, from reading input data in
a tab-deliminated text format to writing the result to a local text file
<code>output.txt</code>. The output column <code>pval</code> contains
the p-values of the test and the column <code>pval.BH</code> contains
the p-values adjusted for multiple testing using the Benjamini-Hochberg
method (Benjamini &amp; Hochberg, 1995).</p>
<p><strong>Two-group beta-binomial test</strong> (Pham et al.,
Bioinformatics 2010)</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">&quot;https://tvpham.github.io/data/example-3groups.txt&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">head</span>(d)</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co">#&gt;    a1  a2  a3  b1  b2  b3  c1  c2</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co">#&gt; 1 624 496 509 414 394 375 325 288</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co">#&gt; 2 615 854 930 341 523 360 359 329</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co">#&gt; 3 553 560 745 819 490 481 480 500</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co">#&gt; 4 525 412 401 354 321 310 258 228</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co">#&gt; 5 484 284 315 268 282 307 270 298</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#&gt; 6 482 348 400 242 365 367  81 118</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co"># compare the first 3 samples against the next three samples</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>out <span class="ot">&lt;-</span> betabinom<span class="sc">::</span><span class="fu">bb.test</span>(d[, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>], </span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a>                          <span class="fu">colSums</span>(d[, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]), </span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a>                          <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;a&quot;</span>, <span class="dv">3</span>), <span class="fu">rep</span>(<span class="st">&quot;b&quot;</span>, <span class="dv">3</span>)))</span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="co">#&gt; Using 11 thread(s) ...</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="co">#&gt; No. of data rows = 1786, no. of groups = 2, no. of samples = 6...</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="co">#&gt; 5%</span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a><span class="co">#&gt; 11%</span></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a><span class="co">#&gt; 18%</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a><span class="co">#&gt; 23%</span></span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a><span class="co">#&gt; 30%</span></span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a><span class="co">#&gt; 35%</span></span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a><span class="co">#&gt; 41%</span></span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a><span class="co">#&gt; 49%</span></span>
<span id="cb2-26"><a href="#cb2-26" tabindex="-1"></a><span class="co">#&gt; 55%</span></span>
<span id="cb2-27"><a href="#cb2-27" tabindex="-1"></a><span class="co">#&gt; 64%</span></span>
<span id="cb2-28"><a href="#cb2-28" tabindex="-1"></a><span class="co">#&gt; 81%</span></span>
<span id="cb2-29"><a href="#cb2-29" tabindex="-1"></a><span class="co">#&gt; 87%</span></span>
<span id="cb2-30"><a href="#cb2-30" tabindex="-1"></a><span class="co">#&gt; 93%</span></span>
<span id="cb2-31"><a href="#cb2-31" tabindex="-1"></a><span class="co">#&gt; 98%</span></span>
<span id="cb2-32"><a href="#cb2-32" tabindex="-1"></a><span class="co">#&gt; Done.</span></span>
<span id="cb2-33"><a href="#cb2-33" tabindex="-1"></a></span>
<span id="cb2-34"><a href="#cb2-34" tabindex="-1"></a>d.norm <span class="ot">&lt;-</span>  betabinom<span class="sc">::</span><span class="fu">normalize</span>(d[, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>])</span>
<span id="cb2-35"><a href="#cb2-35" tabindex="-1"></a></span>
<span id="cb2-36"><a href="#cb2-36" tabindex="-1"></a><span class="fu">write.table</span>(<span class="fu">cbind</span>(d, d.norm, </span>
<span id="cb2-37"><a href="#cb2-37" tabindex="-1"></a>                  <span class="at">fc =</span> betabinom<span class="sc">::</span><span class="fu">fold.change</span>(d.norm[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], d.norm[, <span class="dv">4</span><span class="sc">:</span><span class="dv">6</span>]),</span>
<span id="cb2-38"><a href="#cb2-38" tabindex="-1"></a>                  <span class="at">pval =</span> out<span class="sc">$</span>p.value,</span>
<span id="cb2-39"><a href="#cb2-39" tabindex="-1"></a>                  <span class="at">pval.BH =</span> <span class="fu">p.adjust</span>(out<span class="sc">$</span>p.value, <span class="at">method =</span> <span class="st">&quot;BH&quot;</span>)),</span>
<span id="cb2-40"><a href="#cb2-40" tabindex="-1"></a>            <span class="at">file =</span> <span class="st">&quot;output.txt&quot;</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span></code></pre></div>
<p><strong>Three-group beta-binomial test</strong> (Pham et al.,
Bioinformatics 2010)</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">&quot;https://tvpham.github.io/data/example-3groups.txt&quot;</span>)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="fu">head</span>(d)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co">#&gt;    a1  a2  a3  b1  b2  b3  c1  c2</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co">#&gt; 1 624 496 509 414 394 375 325 288</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co">#&gt; 2 615 854 930 341 523 360 359 329</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co">#&gt; 3 553 560 745 819 490 481 480 500</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co">#&gt; 4 525 412 401 354 321 310 258 228</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="co">#&gt; 5 484 284 315 268 282 307 270 298</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="co">#&gt; 6 482 348 400 242 365 367  81 118</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="co"># compare the first 3 samples, the next three samples, and the last two samples.</span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>out <span class="ot">&lt;-</span> betabinom<span class="sc">::</span><span class="fu">bb.test</span>(d[, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>],</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a>                          <span class="fu">colSums</span>(d[, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]),</span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a>                          <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;a&quot;</span>, <span class="dv">3</span>), <span class="fu">rep</span>(<span class="st">&quot;b&quot;</span>, <span class="dv">3</span>), <span class="fu">rep</span>(<span class="st">&quot;c&quot;</span>, <span class="dv">2</span>)))</span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="co">#&gt; Using 11 thread(s) ...</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a><span class="co">#&gt; No. of data rows = 1786, no. of groups = 3, no. of samples = 8...</span></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a><span class="co">#&gt; 0%</span></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a><span class="co">#&gt; 5%</span></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a><span class="co">#&gt; 10%</span></span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a><span class="co">#&gt; 16%</span></span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a><span class="co">#&gt; 23%</span></span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a><span class="co">#&gt; 33%</span></span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a><span class="co">#&gt; 40%</span></span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a><span class="co">#&gt; 45%</span></span>
<span id="cb3-26"><a href="#cb3-26" tabindex="-1"></a><span class="co">#&gt; 51%</span></span>
<span id="cb3-27"><a href="#cb3-27" tabindex="-1"></a><span class="co">#&gt; 58%</span></span>
<span id="cb3-28"><a href="#cb3-28" tabindex="-1"></a><span class="co">#&gt; 63%</span></span>
<span id="cb3-29"><a href="#cb3-29" tabindex="-1"></a><span class="co">#&gt; 70%</span></span>
<span id="cb3-30"><a href="#cb3-30" tabindex="-1"></a><span class="co">#&gt; 76%</span></span>
<span id="cb3-31"><a href="#cb3-31" tabindex="-1"></a><span class="co">#&gt; 81%</span></span>
<span id="cb3-32"><a href="#cb3-32" tabindex="-1"></a><span class="co">#&gt; 87%</span></span>
<span id="cb3-33"><a href="#cb3-33" tabindex="-1"></a><span class="co">#&gt; 93%</span></span>
<span id="cb3-34"><a href="#cb3-34" tabindex="-1"></a><span class="co">#&gt; 98%</span></span>
<span id="cb3-35"><a href="#cb3-35" tabindex="-1"></a><span class="co">#&gt; Done.</span></span>
<span id="cb3-36"><a href="#cb3-36" tabindex="-1"></a></span>
<span id="cb3-37"><a href="#cb3-37" tabindex="-1"></a>d.norm <span class="ot">&lt;-</span>  betabinom<span class="sc">::</span><span class="fu">normalize</span>(d[, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>])</span>
<span id="cb3-38"><a href="#cb3-38" tabindex="-1"></a></span>
<span id="cb3-39"><a href="#cb3-39" tabindex="-1"></a><span class="fu">write.table</span>(<span class="fu">cbind</span>(d, d.norm, </span>
<span id="cb3-40"><a href="#cb3-40" tabindex="-1"></a>                  <span class="at">pval =</span> out<span class="sc">$</span>p.value,</span>
<span id="cb3-41"><a href="#cb3-41" tabindex="-1"></a>                  <span class="at">pval.BH =</span> <span class="fu">p.adjust</span>(out<span class="sc">$</span>p.value, <span class="at">method =</span> <span class="st">&quot;BH&quot;</span>)),</span>
<span id="cb3-42"><a href="#cb3-42" tabindex="-1"></a>            <span class="at">file =</span> <span class="st">&quot;output.txt&quot;</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span></code></pre></div>
<p><strong>Two-group paired beta-binomial test</strong> (Pham &amp;
Jimenez, Bioinformatics 2012)</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">&quot;https://tvpham.github.io/data/example-paired.txt&quot;</span>)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="fu">head</span>(d)</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#&gt;   pre.1 pre.2 pre.3 post.1 post.2 post.3</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#&gt; 1   575   179   335    505    172    204</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#&gt; 2   294   245   256    396    390    265</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#&gt; 3   293   282   320    372    240    204</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co">#&gt; 4   303   282   250    307    243    227</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co">#&gt; 5   396   271   171    327    216    103</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="co">#&gt; 6   238   261   271    245    234    215</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a>out <span class="ot">&lt;-</span> betabinom<span class="sc">::</span><span class="fu">ibb.test</span>(d[, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>], </span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a>                           <span class="fu">colSums</span>(d[, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]), </span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a>                           <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;pre_treatment&quot;</span>, <span class="dv">3</span>), <span class="fu">rep</span>(<span class="st">&quot;post_treatment&quot;</span>, <span class="dv">3</span>)))</span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a><span class="co">#&gt; Using 11 thread(s) ...</span></span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a><span class="co">#&gt; No. of data rows = 2919, no. of pair(s) = 3...</span></span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a><span class="co">#&gt; 0%</span></span>
<span id="cb4-18"><a href="#cb4-18" tabindex="-1"></a><span class="co">#&gt; 5%</span></span>
<span id="cb4-19"><a href="#cb4-19" tabindex="-1"></a><span class="co">#&gt; 11%</span></span>
<span id="cb4-20"><a href="#cb4-20" tabindex="-1"></a><span class="co">#&gt; 16%</span></span>
<span id="cb4-21"><a href="#cb4-21" tabindex="-1"></a><span class="co">#&gt; 21%</span></span>
<span id="cb4-22"><a href="#cb4-22" tabindex="-1"></a><span class="co">#&gt; 26%</span></span>
<span id="cb4-23"><a href="#cb4-23" tabindex="-1"></a><span class="co">#&gt; 31%</span></span>
<span id="cb4-24"><a href="#cb4-24" tabindex="-1"></a><span class="co">#&gt; 36%</span></span>
<span id="cb4-25"><a href="#cb4-25" tabindex="-1"></a><span class="co">#&gt; 41%</span></span>
<span id="cb4-26"><a href="#cb4-26" tabindex="-1"></a><span class="co">#&gt; 46%</span></span>
<span id="cb4-27"><a href="#cb4-27" tabindex="-1"></a><span class="co">#&gt; 51%</span></span>
<span id="cb4-28"><a href="#cb4-28" tabindex="-1"></a><span class="co">#&gt; 57%</span></span>
<span id="cb4-29"><a href="#cb4-29" tabindex="-1"></a><span class="co">#&gt; 63%</span></span>
<span id="cb4-30"><a href="#cb4-30" tabindex="-1"></a><span class="co">#&gt; 68%</span></span>
<span id="cb4-31"><a href="#cb4-31" tabindex="-1"></a><span class="co">#&gt; 73%</span></span>
<span id="cb4-32"><a href="#cb4-32" tabindex="-1"></a><span class="co">#&gt; 78%</span></span>
<span id="cb4-33"><a href="#cb4-33" tabindex="-1"></a><span class="co">#&gt; 83%</span></span>
<span id="cb4-34"><a href="#cb4-34" tabindex="-1"></a><span class="co">#&gt; 88%</span></span>
<span id="cb4-35"><a href="#cb4-35" tabindex="-1"></a><span class="co">#&gt; 94%</span></span>
<span id="cb4-36"><a href="#cb4-36" tabindex="-1"></a><span class="co">#&gt; 99%</span></span>
<span id="cb4-37"><a href="#cb4-37" tabindex="-1"></a><span class="co">#&gt; Done.</span></span>
<span id="cb4-38"><a href="#cb4-38" tabindex="-1"></a></span>
<span id="cb4-39"><a href="#cb4-39" tabindex="-1"></a>d.norm <span class="ot">&lt;-</span>  betabinom<span class="sc">::</span><span class="fu">normalize</span>(d[, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>])</span>
<span id="cb4-40"><a href="#cb4-40" tabindex="-1"></a></span>
<span id="cb4-41"><a href="#cb4-41" tabindex="-1"></a><span class="fu">write.table</span>(<span class="fu">cbind</span>(d, d.norm, </span>
<span id="cb4-42"><a href="#cb4-42" tabindex="-1"></a>                  <span class="at">fc =</span> out<span class="sc">$</span>fc,</span>
<span id="cb4-43"><a href="#cb4-43" tabindex="-1"></a>                  <span class="at">pval =</span> out<span class="sc">$</span>p.value,</span>
<span id="cb4-44"><a href="#cb4-44" tabindex="-1"></a>                  <span class="at">pval.BH =</span> <span class="fu">p.adjust</span>(out<span class="sc">$</span>p.value, <span class="at">method =</span> <span class="st">&quot;BH&quot;</span>)),</span>
<span id="cb4-45"><a href="#cb4-45" tabindex="-1"></a>            <span class="at">file =</span> <span class="st">&quot;output.txt&quot;</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span></code></pre></div>
</div>
<div id="unpaired-test" class="section level2">
<h2>Unpaired test</h2>
<p>The beta-binomial test (Pham et al., Bioinformatics 2010) is used for
significance analysis of independent samples for two or more groups.
Suppose that a vector <code>x</code> contains the count numbers and a
vector <code>tx</code> the corresponding normalization sizes, for
example the total spectral counts per sample in a proteomics experiment.
In addition, the group information is specified in a vector
<code>group</code>. We perform a beta-binomial test as follows.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">1</span>, <span class="dv">10</span>, <span class="dv">9</span>, <span class="dv">11</span>, <span class="dv">2</span>, <span class="dv">8</span>)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>tx <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">19609</span>, <span class="dv">19053</span>, <span class="dv">19235</span>, <span class="dv">19374</span>, <span class="dv">18868</span>, <span class="dv">19018</span>, <span class="dv">18844</span>, <span class="dv">19271</span>)</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;cancer&quot;</span>, <span class="dv">3</span>), <span class="fu">rep</span>(<span class="st">&quot;normal&quot;</span>, <span class="dv">5</span>))</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>betabinom<span class="sc">::</span><span class="fu">bb.test</span>(x, tx, group)</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="co">#&gt; Using a single CPU core ...</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co">#&gt; No. of data rows = 1, no. of groups = 2, no. of samples = 8...</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co">#&gt; 100%</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co">#&gt; Done.</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co">#&gt; $p.value</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="co">#&gt; [1] 0.01568598</span></span></code></pre></div>
<p>The test can be applied to a data matrix row by row. The following
example compares three groups, the first group consisting of columns 1
to 3 of a data matrix, the second columns 4 to 6, and the third columns
7 and 8.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">&quot;https://tvpham.github.io/data/example-3groups.txt&quot;</span>)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co"># compare 3 groups, using all available CPU cores</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>out <span class="ot">&lt;-</span> betabinom<span class="sc">::</span><span class="fu">bb.test</span>(d[, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>], </span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>                          <span class="fu">colSums</span>(d[, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]), </span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>                          <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;a&quot;</span>, <span class="dv">3</span>), <span class="fu">rep</span>(<span class="st">&quot;b&quot;</span>, <span class="dv">3</span>), <span class="fu">rep</span>(<span class="st">&quot;c&quot;</span>, <span class="dv">2</span>)))</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#&gt; Using 11 thread(s) ...</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co">#&gt; No. of data rows = 1786, no. of groups = 3, no. of samples = 8...</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="co">#&gt; 0%</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co">#&gt; 5%</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="co">#&gt; 11%</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="co">#&gt; 16%</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="co">#&gt; 21%</span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="co">#&gt; 28%</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="co">#&gt; 34%</span></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="co">#&gt; 40%</span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="co">#&gt; 46%</span></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a><span class="co">#&gt; 52%</span></span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a><span class="co">#&gt; 58%</span></span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a><span class="co">#&gt; 77%</span></span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a><span class="co">#&gt; 83%</span></span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a><span class="co">#&gt; 88%</span></span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a><span class="co">#&gt; 93%</span></span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a><span class="co">#&gt; 98%</span></span>
<span id="cb6-25"><a href="#cb6-25" tabindex="-1"></a><span class="co">#&gt; Done.</span></span></code></pre></div>
<p>The output <code>out$p.value</code> contains p-values for each row of
<code>d</code>. For multiple testing correction, use the
<code>p.adjust</code> function in R. For example, to apply the Benjamini
&amp; Hochberg method for p-value adjustment</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>pval.BH <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(out<span class="sc">$</span>p.value, <span class="at">method =</span> <span class="st">&quot;BH&quot;</span>)</span></code></pre></div>
<div id="data-normalization" class="section level3">
<h3>Data normalization</h3>
<p>The beta-binomial test takes <strong>raw counts as input</strong> and
normalizes the values internally. To obtain normalized values, we
multiply each sample by a scaling factor so that the total counts are
equal across samples. This procedure is implemented in the
<code>normalize</code> function.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>d.norm <span class="ot">&lt;-</span>  betabinom<span class="sc">::</span><span class="fu">normalize</span>(d[, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>])</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="co"># check -- all values should be equal</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="fu">colSums</span>(d.norm)</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co">#&gt;    a1    a2    a3    b1    b2    b3    c1    c2 </span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co">#&gt; 19159 19159 19159 19159 19159 19159 19159 19159</span></span></code></pre></div>
<p>The matrix <code>d.norm</code> contains the normalized data. We can
combine the normalized data and the result of the beta-binomial test by
<code>cbind(d.norm, out$p.value)</code>, and subsequently save or view
the resulting matrix</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">cbind</span>(d.norm, out<span class="sc">$</span>p.value))</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="co">#&gt;         a1       a2       a3       b1       b2       b3        c1       c2</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="co">#&gt; 1 609.6800 498.7595 506.9889 409.4057 400.0766 377.7803 330.43276 286.3262</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="co">#&gt; 2 600.8866 858.7512 926.3254 337.2158 531.0662 362.6691 365.00111 327.0879</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="co">#&gt; 3 540.3094 563.1155 742.0564 809.9113 497.5572 484.5661 488.02377 497.0941</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="co">#&gt; 4 512.9520 414.2921 399.4156 350.0715 325.9508 312.2983 262.31278 226.6749</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="co">#&gt; 5 472.8929 285.5800 313.7554 265.0259 286.3493 309.2761 274.51337 296.2681</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="co">#&gt; 6 470.9388 349.9361 398.4195 239.3144 370.6294 369.7209  82.35401 117.3142</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="co">#&gt;    out$p.value</span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="co">#&gt; 1 0.0001164498</span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="co">#&gt; 2 0.0016150240</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a><span class="co">#&gt; 3 0.4501729545</span></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a><span class="co">#&gt; 4 0.0003105657</span></span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a><span class="co">#&gt; 5 0.2521494212</span></span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a><span class="co">#&gt; 6 0.0001057092</span></span></code></pre></div>
</div>
<div id="fold-change-calculation" class="section level3">
<h3>Fold change calculation</h3>
<p>The package provides a convenient function to calculate the ratio
between the averages of two sample groups row by row. For example, to
calculate the fold change between all group pairs</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>fc12 <span class="ot">&lt;-</span> betabinom<span class="sc">::</span><span class="fu">fold.change</span>(d.norm[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], d.norm[, <span class="dv">4</span><span class="sc">:</span><span class="dv">6</span>])</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>fc13 <span class="ot">&lt;-</span> betabinom<span class="sc">::</span><span class="fu">fold.change</span>(d.norm[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], d.norm[, <span class="dv">7</span><span class="sc">:</span><span class="dv">8</span>])</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>fc23 <span class="ot">&lt;-</span> betabinom<span class="sc">::</span><span class="fu">fold.change</span>(d.norm[, <span class="dv">4</span><span class="sc">:</span><span class="dv">6</span>], d.norm[, <span class="dv">7</span><span class="sc">:</span><span class="dv">8</span>], <span class="at">BIG =</span> <span class="dv">100</span>)</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">cbind</span>(d.norm, out<span class="sc">$</span>p.value, fc12, fc13, fc23))</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a><span class="co">#&gt;         a1       a2       a3       b1       b2       b3        c1       c2</span></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a><span class="co">#&gt; 1 609.6800 498.7595 506.9889 409.4057 400.0766 377.7803 330.43276 286.3262</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a><span class="co">#&gt; 2 600.8866 858.7512 926.3254 337.2158 531.0662 362.6691 365.00111 327.0879</span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a><span class="co">#&gt; 3 540.3094 563.1155 742.0564 809.9113 497.5572 484.5661 488.02377 497.0941</span></span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a><span class="co">#&gt; 4 512.9520 414.2921 399.4156 350.0715 325.9508 312.2983 262.31278 226.6749</span></span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a><span class="co">#&gt; 5 472.8929 285.5800 313.7554 265.0259 286.3493 309.2761 274.51337 296.2681</span></span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a><span class="co">#&gt; 6 470.9388 349.9361 398.4195 239.3144 370.6294 369.7209  82.35401 117.3142</span></span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a><span class="co">#&gt;    out$p.value      fc12      fc13      fc23</span></span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a><span class="co">#&gt; 1 0.0001164498 -1.360633 -1.746148 -1.283335</span></span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a><span class="co">#&gt; 2 0.0016150240 -1.938309 -2.298320 -1.185735</span></span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a><span class="co">#&gt; 3 0.4501729545 -1.029825 -1.248907 -1.212738</span></span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a><span class="co">#&gt; 4 0.0003105657 -1.342337 -1.808716 -1.347438</span></span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a><span class="co">#&gt; 5 0.2521494212 -1.245834 -1.252351 -1.005232</span></span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a><span class="co">#&gt; 6 0.0001057092 -1.244604 -4.071068 -3.270976</span></span></code></pre></div>
<p>Note that a positive fold change value means up-regulation where the
average of the second group is higher than that of the first group.
Conversely, a negative value means down-regulation where the average of
the first group is higher than that of the second. If one group contains
all zeros, a positive or negative big value is returned (default BIG =
10000).</p>
</div>
</div>
<div id="paired-test" class="section level2">
<h2>Paired test</h2>
<p>The inverted beta-binomial test (Pham &amp; Jimenez, Bioinformatics
2012) is used for paired sample testing, for example between
pre-treatment and post-treatment data. The following is an example of a
paired test.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">33</span>, <span class="dv">32</span>, <span class="dv">86</span>, <span class="dv">51</span>, <span class="dv">52</span>, <span class="dv">149</span>)</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>tx <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">7742608</span>, <span class="dv">15581382</span>, <span class="dv">20933491</span>, <span class="dv">7126839</span>, <span class="dv">13842297</span>, <span class="dv">14760103</span>)</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;cancer&quot;</span>, <span class="dv">3</span>), <span class="fu">rep</span>(<span class="st">&quot;normal&quot;</span>, <span class="dv">3</span>))</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>betabinom<span class="sc">::</span><span class="fu">ibb.test</span>(x, tx, group)</span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a><span class="co">#&gt; Using a single CPU core ...</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a><span class="co">#&gt; No. of data rows = 1, no. of pair(s) = 3...</span></span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a><span class="co">#&gt; 100%</span></span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a><span class="co">#&gt; Done.</span></span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a><span class="co">#&gt; $p.value</span></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a><span class="co">#&gt; [1] 0.004103636</span></span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a><span class="co">#&gt; $fc</span></span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a><span class="co">#&gt; [1] 2.137632</span></span></code></pre></div>
<p>Analogously to the unpaired situation, the test can be performed for
a data matrix row by row. The following example compares two groups
where columns 1 to 3 are respectively paired with columns 4 to 6.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">&quot;https://tvpham.github.io/data/example-paired.txt&quot;</span>)</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="co"># perform a paired test for all rows</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>out <span class="ot">&lt;-</span> betabinom<span class="sc">::</span><span class="fu">ibb.test</span>(d[, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>], </span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>                           <span class="fu">colSums</span>(d[, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]), </span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>                           <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;pre_treatment&quot;</span>, <span class="dv">3</span>), <span class="fu">rep</span>(<span class="st">&quot;post_treatment&quot;</span>, <span class="dv">3</span>)))</span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="co">#&gt; Using 11 thread(s) ...</span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="co">#&gt; No. of data rows = 2919, no. of pair(s) = 3...</span></span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="co">#&gt; 0%</span></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a><span class="co">#&gt; 5%</span></span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a><span class="co">#&gt; 10%</span></span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a><span class="co">#&gt; 16%</span></span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a><span class="co">#&gt; 21%</span></span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a><span class="co">#&gt; 26%</span></span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a><span class="co">#&gt; 31%</span></span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a><span class="co">#&gt; 37%</span></span>
<span id="cb12-17"><a href="#cb12-17" tabindex="-1"></a><span class="co">#&gt; 42%</span></span>
<span id="cb12-18"><a href="#cb12-18" tabindex="-1"></a><span class="co">#&gt; 47%</span></span>
<span id="cb12-19"><a href="#cb12-19" tabindex="-1"></a><span class="co">#&gt; 52%</span></span>
<span id="cb12-20"><a href="#cb12-20" tabindex="-1"></a><span class="co">#&gt; 57%</span></span>
<span id="cb12-21"><a href="#cb12-21" tabindex="-1"></a><span class="co">#&gt; 62%</span></span>
<span id="cb12-22"><a href="#cb12-22" tabindex="-1"></a><span class="co">#&gt; 67%</span></span>
<span id="cb12-23"><a href="#cb12-23" tabindex="-1"></a><span class="co">#&gt; 73%</span></span>
<span id="cb12-24"><a href="#cb12-24" tabindex="-1"></a><span class="co">#&gt; 78%</span></span>
<span id="cb12-25"><a href="#cb12-25" tabindex="-1"></a><span class="co">#&gt; 83%</span></span>
<span id="cb12-26"><a href="#cb12-26" tabindex="-1"></a><span class="co">#&gt; 88%</span></span>
<span id="cb12-27"><a href="#cb12-27" tabindex="-1"></a><span class="co">#&gt; 93%</span></span>
<span id="cb12-28"><a href="#cb12-28" tabindex="-1"></a><span class="co">#&gt; 98%</span></span>
<span id="cb12-29"><a href="#cb12-29" tabindex="-1"></a><span class="co">#&gt; Done.</span></span></code></pre></div>
<p>The result <code>out</code> is a list of two elements where
<code>p.value</code> is the p-value of the test and <code>fc</code> an
estimation of the common fold change. We can calculate the normalized
data and write out the result as follows.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>d.norm <span class="ot">&lt;-</span> betabinom<span class="sc">::</span><span class="fu">normalize</span>(d[, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>])</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">cbind</span>(d.norm, out<span class="sc">$</span>p.value, out<span class="sc">$</span>fc))</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="co">#&gt;      pre.1    pre.2    pre.3   post.1   post.2   post.3 out$p.value    out$fc</span></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="co">#&gt; 1 594.2861 176.8823 347.1786 490.0689 164.2960 208.5462 0.064856368 -1.297663</span></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a><span class="co">#&gt; 2 303.8611 242.1015 265.3067 384.2916 372.5316 270.9056 0.072525346  1.259570</span></span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a><span class="co">#&gt; 3 302.8275 278.6637 331.6333 361.0012 229.2502 208.5462 0.334902070 -1.168634</span></span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a><span class="co">#&gt; 4 313.1629 278.6637 259.0885 297.9231 232.1159 232.0588 0.045979358 -1.117328</span></span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a><span class="co">#&gt; 5 409.2823 267.7939 177.2166 317.3317 206.3252 105.2954 0.006665643 -1.356184</span></span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a><span class="co">#&gt; 6 245.9828 257.9122 280.8520 237.7562 223.5190 219.7913 0.048216659 -1.151104</span></span></code></pre></div>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<ol style="list-style-type: decimal">
<li><p>Benjamini Y, Hochberg Y (1995) Controlling the false discovery
rate: a practical and powerful approach to multiple testing. <em>Journal
of the Royal Statistical Society Series B</em> 57, 289â€“300.</p></li>
<li><p>Pham TV, Piersma SR, Warmoes M, Jimenez CR (2010) On the beta
binomial model for analysis of spectral count data in label-free tandem
mass spectrometry-based proteomics. <em>Bioinformatics</em>,
26(3):363-369.</p></li>
<li><p>Pham TV, Jimenez CR (2012) An accurate paired sample test for
count data. <em>Bioinformatics</em>, 28(18):i596-i602.</p></li>
</ol>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
